chore: adds apdex telemetry details

latin-panda opened the pull request:
<!--
Please use semantic PR titles that respect this format:

<type>(#<issue number>): <subject>

Quick example:

feat(#1234): add hat wobble
^--^(#^--^): ^------------^
|     |      |
|     |      + - > subject
|     |
|     + -------- > issue number
|
+ -------------- > type: chore, feat, fix, perf.

https://docs.communityhealthtoolkit.org/contribute/code/workflow/#commit-message-format
-->

# Description
https://github.com/medic/cht-core/issues/8821
[CHT CORE PR](https://github.com/medic/cht-core/pull/8822)

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

Benmuiruri commented on content/en/apps/guides/performance/telemetry.md:
```diff
@@ -36,6 +36,7 @@ The telemetry data gathered changes with different versions of the framework. Cu
 
 | Field | Description |
 |----|----|
+| `<page>:<component>:apdex:<apdex_level>` | The time taken to load the page in the Apdex level. The tolerance level is `satisfied` when the duration is less or equal than 3s; `tolerable` when the duration is more than 3s but less or equal than 12s; `frustrated` when duration is more than 12s. Added in 4.6. |
```
```suggestion
| `<page>:<component>:apdex:<apdex_level>` | The time taken to load the page, categorized by Apdex level. The tolerance level is `satisfied` when the duration is less than or equal to 3s; `tolerable` when the duration is more than 3s but less than or equal to 12s; `frustrated` when duration is more than 12s. Added in 4.6. |
```

latin-panda replied:
Thank you!

Benmuiruri reviewed (approved):
I would suggest that small grammar change

Benmuiruri commented on content/en/apps/guides/performance/telemetry.md:
```diff
@@ -36,6 +36,7 @@ The telemetry data gathered changes with different versions of the framework. Cu
 
 | Field | Description |
 |----|----|
+| `<page>:<component>:apdex:<apdex_level>` | The time taken to load the page, categorized by Apdex level. The Apdex leve is `satisfied` when the duration is less than or equal to 3s; `tolerable` when the duration is more than 3s but less than or equal to 12s; `frustrated` when duration is more than 12s. Added in 4.6. |
```
```suggestion
| `<page>:<component>:apdex:<apdex_level>` | The time taken to load the page, categorized by Apdex level. The Apdex level is `satisfied` when the duration is less than or equal to 3s; `tolerable` when the duration is more than 3s but less than or equal to 12s; `frustrated` when duration is more than 12s. Added in 4.6. |
```

latin-panda commented:
@michaelkohn are these updates clear?

michaelkohn reviewed (changes requested):
I've left some comments and suggestions that I don't know are actually accurate, but helps get across what I think would be helpful to see. If I don't ask / we don't address them here, I'm going to be asking about them in the forum at some point ;-).   Overall these are great additions that I'm excited about!

latin-panda commented on content/en/apps/guides/performance/telemetry.md:
```diff
@@ -36,6 +36,7 @@ The telemetry data gathered changes with different versions of the framework. Cu
 
 | Field | Description |
 |----|----|
+| `<page>:<component>:apdex:<apdex_level>` | The time taken to load the page, categorized by Apdex level. The Apdex leve is `satisfied` when the duration is less than or equal to 3s; `tolerable` when the duration is more than 3s but less than or equal to 12s; `frustrated` when duration is more than 12s. Added in 4.6. |
```
I'm going to put all of these in separate lines. It will be repetitive, but it will be explicit in terms of which Apdex we have.

latin-panda replied:
All the apdex records are together so it's easier to find what's covered by apdex

latin-panda commented on content/en/apps/guides/performance/telemetry.md:
```diff
@@ -48,18 +49,27 @@ The telemetry data gathered changes with different versions of the framework. Cu
-| `enketo:reports:<form>:<action>:<component>` | The time taken to fill in Enketo forms. The `action` can either be "add" or "edit". The `component` is one of: "render" covers getting the form and rendering it on screen; "user_edit_time" is the time the user took to fill in and submit the form; or "save" is about converting the form into a report and saving it. |
-| `enketo:contacts:<form>:<action>:<component>` | As above but for Contact forms. The `action` can either be "add" or "edit". |
+| `contact_list:load` | The time taken to load contact list. Added in 4.6. |
+| `contact_list:query` | The time taken to query contacts. This includes free text search and sort. Added in 4.6.  |
```
> It's not clear to me if this is talking only about the search bar on the contacts tab or if this is a subset of contact_list:load

It's both. When the contact list loads (`contact_list:load`) it does a query (`contact_list:query`) to bring the data. And when the CHW searches by a term or do a sort, the same query process (`contact_list:query`) is recorded. ([The code](https://github.com/medic/cht-core/blob/900218e9b4594c11e01809d1f1333cd955e7d310/webapp/src/ts/modules/contacts/contacts.component.ts#L313), for the technical eyes)

> What are all the different user interactions that will record against (ie count towards) this telemetry key?

The initial load of Contact List when visiting the tab, search and sort. They are all the same process. The duration recorded includes:
- Build the search criteria 
- Search (here it records `search:contacts:<filter[:filter]>`)
- Once the search has finished, update the home place 
- Format the search results so it can be displayed - includes calculation of visits and overdue if enabled.
- Update the scroll in the view 
- Update FAB
- Send for display
After all this, the query process is considered finished and the Apdex is calculated (last line of the function).

Not sure how to explain all this in the docs 😅

latin-panda replied:
I managed to add a bit of detail to make it too long, I definitely need feedback about how to improve it

latin-panda commented on content/en/apps/guides/performance/telemetry.md:
```diff
@@ -48,18 +49,27 @@ The telemetry data gathered changes with different versions of the framework. Cu
 | `tasks:load` | The time taken to load the tasks page. Added in 3.9 | 
 | `tasks:refresh` | The time taken to refresh tasks on the tasks page. Added in 3.9 | 
+| `report_list:load` | The time taken to load the report list. Added in 4.6. |
+| `report_list:query` | The time taken to query reports. This includes free text search and filters. Added in 4.6 |
```
Following the same line as [this answer](https://github.com/medic/cht-docs/pull/1278#discussion_r1475636136)

This is initial load of report list, search, filters, they all uses the same query process (`report_list:query`)

A bit simpler, the duration recorded includes:

- Build the search criteria
- Search (here it records search:reports:<filter[:filter]>)
- Update the scroll in the view
- Send for display (The FAB don't change)

After all this, the query process is considered finished and the Apdex is calculated (last line of the function).
This only covers the reports tab scope, different from search:reports:<filter[:filter]>, that is triggered in other places as you mentioned the other day.

latin-panda replied:
I added a bit more detail. I need feedback on how to improve it :)

latin-panda commented:
@Benmuiruri @michaelkohn please review the documentation about Apdex and new telemetry records

michaelkohn commented:
Thanks @latin-panda .  Right now we are pretty much duplicating telemetry keys in the apdex section. There are currently ~18 duplicates and I think it's actually missing all the granular ones.  What do you think about adding a column in the main table specifying whether or not that field records apdex and then just having the explanation and 1 example below in the apdex section?

latin-panda commented:
> Right now we are pretty much duplicating telemetry keys in the apdex section. There are currently ~18 duplicates and I think it's actually missing all the granular ones.

@michaelkohn The new [granular telemetry is in the first table.](https://github.com/medic/cht-docs/pull/1278/files#diff-4c4fda2d298052191547f4f0e7d5393d1cdaa13908d917490e358eef77d295e0R62) The granular are "pieces" of a bigger process (load contact details), therefore we don't calculate the Apdex for these: 

```
contact_detail:<_form>:load:contact_data
contact_detail:<_form>:load:load_descendants
contact_detail:<_form>:load:load_reports
contact_detail:<_form>:load:load_targets
contact_detail:<_form>:load:load_tasks
contact_detail:<_form>:load:load_contact_summary
```

latin-panda commented:
@michaelkohn and @Benmuiruri, I have applied the feedback where we removed the Apdex table and just added a column to the previous telemetry table. 
Please review

Benmuiruri commented on content/en/apps/guides/performance/telemetry.md:
```diff
@@ -22,76 +22,95 @@ Versions prior to v3.12.0 do aggregate in a monthly basis instead of daily, and
+| `sumsqr` | The sum of squares of the times recorded in milliseconds. This is useful to see the variance, for example, a lower sumsqr can be indicative of having data closer together. |
 
-The telemetry data gathered changes with different versions of the framework. Currently, the data points collected are:
+The CHT records the Apdex (Application Performance Index) that is an open standard for measuring performance of software applications. Its purpose is to convert measurements into insights about user satisfaction, by specifying an uniform way to analyze and report on the degree to which measured performance meets user expectations. 
```
```suggestion
The CHT records the Apdex (Application Performance Index) that is an open standard for measuring performance of software applications. Its purpose is to convert measurements into insights about user satisfaction, by specifying a uniform way to analyze and report on the degree to which measured performance meets user expectations. 
```

latin-panda replied:
Thanks!

latin-panda commented on content/en/apps/guides/performance/telemetry.md:
```diff
@@ -22,76 +22,95 @@ Versions prior to v3.12.0 do aggregate in a monthly basis instead of daily, and
+| `boot_time` | The overall boot time including loading the code, purging, and accessing the database. | Yes. Added in 4.7|
+| `boot_time:1:to_first_code_execution` | The time between the page loading and the JavaScript starting to run. | |
+| `boot_time:2:to_bootstrap` | The time between JavaScript starting and the bootstrapping (purging, initial replication, etc) to complete. | |
+| `boot_time:2_1:to_replication` | The time it took to complete initial replication. If initial replication was interrupted and retried, this value will be incorrect. Added in 3.14. | |
```
Thanks!
